packages/contracts/src/zksync/PluginSetupProcessorUpgradeable.sol

```diff
@@ -0,0 +1,753 @@
+                    initData: _initData
+                });
+            }
+        } else {
+            try PluginUUPSUpgradeable(_proxy).upgradeTo(_implementation) {} catch Error(
+                string memory reason
+            ) {
+                revert(reason);
+            } catch (bytes memory /*lowLevelData*/) {
+                revert PluginProxyUpgradeFailed({
+                    proxy: _proxy,
+                    implementation: _implementation,
+                    initData: _initData
+                });
+            }
+        }
+    }
+
+    /// @notice Checks if a caller can apply a setup. The caller can be either the DAO to which the plugin setup is applied to or another account to which the DAO has granted the respective permission.
+    /// @param _dao The address of the applying DAO.
+    /// @param _permissionId The permission ID.
+    function _canApply(address _dao, bytes32 _permissionId) private view {
+        if (
+            msg.sender != _dao &&
+            !DAO(payable(_dao)).hasPermission(address(this), msg.sender, _permissionId, bytes(""))
+        ) {
+            revert SetupApplicationUnauthorized({
+                dao: _dao,
+                caller: msg.sender,
+                permissionId: _permissionId
+            });
+        }
+    }
+
+    /// @notice A helper to emit the `UpdatePrepared` event from the supplied, structured data.
+    /// @param _dao The address of the updating DAO.
+    /// @param _preparedSetupId The prepared setup ID.
+    /// @param _params The struct containing the parameters for the `prepareUpdate` function.
+    /// @param _preparedSetupData The deployed plugin's relevant data which consists of helpers and permissions.
+    /// @param _initData The initialization data to be passed to upgradeable contracts when the update is applied
+    /// @dev This functions exists to avoid stack-too-deep errors.
+    function emitPrepareUpdateEvent(
+        address _dao,
+        bytes32 _preparedSetupId,
+        PrepareUpdateParams calldata _params,
+        IPluginSetup.PreparedSetupData memory _preparedSetupData,
+        bytes memory _initData
+    ) private {
+        emit UpdatePrepared({
+            sender: msg.sender,
+            dao: _dao,
+            preparedSetupId: _preparedSetupId,
+            pluginSetupRepo: _params.pluginSetupRepo,
+            versionTag: _params.newVersionTag,
+            setupPayload: _params.setupPayload,
+            preparedSetupData: _preparedSetupData,
+            initData: _initData
+        });
+    }
+
+    /// @notice Internal method authorizing the upgrade of the contract via the [upgradeability mechanism for UUPS proxies](https://docs.openzeppelin.com/contracts/4.x/api/proxy#UUPSUpgradeable) (see [ERC-1822](https://eips.ethereum.org/EIPS/eip-1822)).
+    /// @dev The caller must have the `UPGRADE_PSP_PERMISSION_ID ` permission.
+    function _authorizeUpgrade(address) internal virtual override auth(UPGRADE_PSP_PERMISSION_ID) {}
+}
```

Comment:
Should we add a storage gap here? I mean I don't expect any contract to inherit from this one, but it doesn't hurt.